From 621123f2656b33b7b83f2f5322f909601369f29c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Yao=20Wei=20=28=E9=AD=8F=E9=8A=98=E5=BB=B7=29?=
 <yao.wei@canonical.com>
Date: Thu, 16 Feb 2023 14:35:27 +0800
Subject: [PATCH 1/2] xkb: add XkbForceUpdateDeviceLEDs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This function is to force update LED states to the hardware from the
device state.  Used if the actual LED state was externally set and need
to push current state to the hardware e.g. switching between VTs.

Signed-off-by: Yao Wei (魏銘廷) <yao.wei@canonical.com>
---
 include/xkbsrv.h |  2 ++
 xkb/xkbLEDs.c    | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 41 insertions(+)

diff --git a/include/xkbsrv.h b/include/xkbsrv.h
index 5b4f71f76..21cd8768d 100644
--- a/include/xkbsrv.h
+++ b/include/xkbsrv.h
@@ -505,6 +505,8 @@ extern _X_EXPORT void XkbUpdateIndicators(DeviceIntPtr /* keybd */ ,
                                           XkbEventCausePtr      /* cause */
     );
 
+extern void XkbForceUpdateDeviceLEDs(DeviceIntPtr /* keybd */);
+
 extern _X_EXPORT void XkbUpdateAllDeviceIndicators(XkbChangesPtr /* changes */,
                                                    XkbEventCausePtr /* cause */
     );
diff --git a/xkb/xkbLEDs.c b/xkb/xkbLEDs.c
index d4690dad9..2fefc99a5 100644
--- a/xkb/xkbLEDs.c
+++ b/xkb/xkbLEDs.c
@@ -434,6 +434,45 @@ XkbUpdateIndicators(DeviceIntPtr dev,
 
 /***====================================================================***/
 
+        /*
+         * void
+         * XkbForceUpdateDeviceLEDs(DeviceIntPtr dev)
+         *
+         * Force update LED states to the hardware from the device state
+         * specified by 'dev'.
+         *
+         * If 'dev' is a master device, this function will also force update
+         * its slave devices.
+         *
+         * Used if the actual LED state was externally set and need to push
+         * current state to the hardware e.g. switching between VTs.
+         */
+
+void
+XkbForceUpdateDeviceLEDs(DeviceIntPtr dev)
+{
+    DeviceIntPtr master;
+    XkbSrvLedInfoPtr sli;
+
+    if (!dev->key)
+        return;
+
+    sli = XkbFindSrvLedInfo(dev, XkbDfltXIClass, XkbDfltXIId, 0);
+    XkbDDXUpdateDeviceIndicators(dev, sli, sli->effectiveState);
+
+    if (IsMaster(dev)) {
+        master = dev;
+        nt_list_for_each_entry(dev, inputInfo.devices, next) {
+            if (!dev->key || GetMaster(dev, MASTER_KEYBOARD) != master)
+                continue;
+
+            sli = XkbFindSrvLedInfo(dev, XkbDfltXIClass, XkbDfltXIId, 0);
+            XkbDDXUpdateDeviceIndicators(dev, sli, sli->effectiveState);
+        }
+    }
+    return;
+}
+
 /***====================================================================***/
 
         /*
-- 
GitLab


From 04f8e840192b9d8d1d087d88970a362fb2372b6e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Yao=20Wei=20=28=E9=AD=8F=E9=8A=98=E5=BB=B7=29?=
 <yao.wei@canonical.com>
Date: Thu, 16 Feb 2023 14:36:38 +0800
Subject: [PATCH 2/2] dix: Force update LEDs after device state update in
 EnableDevice
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is to make sure the hardware gets the device states regardless
whether the internal state has changed or not, to overcome situations
that device LEDs are out of sync e.g. switching between VTs.

Signed-off-by: Yao Wei (魏銘廷) <yao.wei@canonical.com>
---
 dix/devices.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dix/devices.c b/dix/devices.c
index f5ab17352..7150734a5 100644
--- a/dix/devices.c
+++ b/dix/devices.c
@@ -422,6 +422,10 @@ EnableDevice(DeviceIntPtr dev, BOOL sendevent)
 
     if (!IsMaster(dev) && !IsFloating(dev))
         XkbPushLockedStateToSlaves(GetMaster(dev, MASTER_KEYBOARD), 0, 0);
+
+    /* Now make sure our LEDs are in sync with the locked state */
+    XkbForceUpdateDeviceLEDs(dev);
+
     RecalculateMasterButtons(dev);
 
     /* initialise an idle timer for this device*/
-- 
GitLab

